software/firmware/europi_script.py

```diff
@@ -52,23 +121,99 @@ def display_name(cls):
+        .. note::
+            Be mindful of how often `save_state_bytes()` is called because
+            writing to disk too often can slow down the performance of your
+            script. Only call save state when state has changed and consider
+            adding a time since last save check to reduce save frequency.
+        """
+        return self._save_state(state, mode='wb')
+
+    def save_state_json(self, state: dict):
+        """Take state as a dict and save as a json string.
+
+        .. note::
+            Be mindful of how often `save_state_json()` is called because
+            writing to disk too often can slow down the performance of your
+            script. Only call save state when state has changed and consider
+            adding a time since last save check to reduce save frequency.
+        """
+        json_str = json.dumps(state)
+        return self._save_state(json_str)
+
+    def _save_state(self, state: str, mode: str ='w'):
+        with open(self._state_filename, mode) as file:
             file.write(state)
+        self._last_saved = ticks_ms()
 
-    def _load_state(self) -> str:
+    def load_state_str(self) -> str:
+        """Check disk for saved state, if it exists, return the raw state value as a string.
+        
+        Check for a previously saved state. If it exists, return state as a
+        string. If no state is found, an empty string will be returned.
+        """
+        return self._load_state()
+
+    def load_state_bytes(self) -> bytes:
+        """Check disk for saved state, if it exists, return the raw state value as bytes.
+        
+        Check for a previously saved state. If it exists, return state as a
+        byte string. If no state is found, an empty string will be returned.
+        """
+        return self._load_state(mode="rb")
+
+    def load_state_json(self) -> dict:
+        """Load previously saved state as a dict.
+
+        Check for a previously saved state. If it exists, return state as a
+        dict. If no state is found, an empty dictionary will be returned.
+        """
+        json_str = self._load_state()
+        if json_str == "":
+            return {}
         try:
-            with open(self._state_filename, 'r') as file:
+            return json.loads(json_str)
+        except ValueError as e:
+            print(f"Unable to decode {json_str}: {e}")
+            return {}
+
+    def _load_state(self, mode: str ='r') -> any:
+        try:
+            with open(self._state_filename, mode) as file:
                 return file.read()
         except OSError as e:
             return ""
-    
-    def _reset_state(self):  # TODO this name clashes with europi.reset_state()
-        try: 
+
+    def remove_state(self):
+        """Remove the state file for this script."""
+        try:
             os.remove(self._state_filename)
         except OSError:
             pass
+    
+    def last_saved(self):
+        """Return the ticks since last save."""
```

Comment:
I might add a note that ticks is measured in milliseconds, as this might not be familiar to some people (although if you're using state save functionality you probably will know what it means...) just a suggestion!